libs/pavex_session_redis/src/lib.rs

```diff
@@ -0,0 +1,327 @@
+    async fn update_ttl(
+        &self,
+        id: &SessionId,
+        ttl: std::time::Duration,
+    ) -> Result<(), UpdateTtlError> {
+        let mut conn = self.0.clone();
+        match redis::pipe()
+            .cmd("EXISTS")
+            .arg(redis_key(id))
+            .cmd("EXPIRE")
+            .arg(redis_key(id))
+            .arg(ttl.as_secs())
+            .query_async(&mut conn)
+            .await
+            .map_err(|e| UpdateTtlError::Other(e.into()))?
+        {
+            (1, 1) => Ok(()),
+            (0, 0) => Err(err_unknown(id).into()),
+            (1, 0) => Err(UpdateTtlError::Other(anyhow::anyhow!(
+                "Session key exists but redis failed to update TTL"
+            ))),
+            (0, 1) => Err(UpdateTtlError::Other(anyhow::anyhow!(
+                "Unexpected reply from redis: redis should not report succesfully setting ttl for non-existent key"
+            ))),
+            (v, w) => Err(UpdateTtlError::Other(anyhow::anyhow!(
+                "Unexpected reply from redis: EXISTS and EXPIRE only return 0 or 1. EXIST returned {:?}, EXPIRE returned {:?}",
+                v,
+                w
+            ))),
+        }
+    }
+
+    /// Loads an existing session record from the store using the provided ID.
+    ///
+    /// If a session with the given ID exists, it is returned. If the session
+    /// does not exist or has been invalidated (e.g., expired), `None` is
+    /// returned.
+    #[tracing::instrument(name = "Load server-side session record", level = tracing::Level::INFO, skip_all)]
+    async fn load(&self, session_id: &SessionId) -> Result<Option<SessionRecord>, LoadError> {
+        let mut conn = self.0.clone();
+        let k = redis_key(session_id);
+        let (ttl_reply, get_reply): (Value, Value) = redis::pipe()
+            .cmd("TTL")
+            .arg(k)
+            .cmd("GET")
+            .arg(k)
+            .query_async(&mut conn)
+            .await
+            .map_err(|e| LoadError::Other(e.into()))?;
+
+        let ttl = match ttl_reply {
+            Value::Int(s) if s >= 0 => std::time::Duration::from_secs(s as u64),
+            Value::Int(-1) => {
+                panic!("Fatal session management error: no TTL set for this session.")
+            }
+            Value::Int(-2) => return Ok(None),
+            _ => {
+                return Err(LoadError::Other(anyhow::anyhow!(
+                    "Redis TTL returned {:?}. Expected integer >= -2",
+                    ttl_reply
+                )));
+            }
+        };
+
+        let state = match get_reply {
+            Value::BulkString(raw) => serde_json::from_slice(&raw)
+                .context("Failed to deserialize the retrieved session state")
+                .map_err(LoadError::DeserializationError)?,
+            _ => {
+                return Err(LoadError::Other(anyhow::anyhow!(
+                    "Redis GET replied {:?}. Expected bulk string.",
```

Comment:
This _may_ leak the stored data, which might contain sensitive information.
We should limit ourselves to stating which value type we got, rather than logging the value itself.